Build system: add and use the distclean target to all makefiles

shindere opened the pull request:
This PR consists in a succession of short and easy to reviw commits.

It systematically adds a distclean target to all the makefiles that have
a clean one and makes sure the root makefile uses it to remove the files
in each directory, rather than removing them directly as was done before.

It is likely that the declaration of distclean could have been factorized
in Makefile.common rather than being added explicitly to each file
but that would have been different of how it's done for other targets
so I preferred not to do this here. The aim of this PR is really
only to make sure every makefile has a distclean target.

Factorizing how the rules are declared can be done later in a dedicated
PR if one wishes so.

shindere committed:
Add a distclean target to ocamltest/Makefile

shindere committed:
Add a distclean target to tools/Makefile

shindere committed:
Add the distclean target to yacc/Makefile

shindere committed:
Add the distclean target to debugger/Makefile

shindere committed:
Add the distclean target to ocamldoc/Makefile

shindere committed:
Add the distclean target to lex/Makefile

shindere committed:
Add the distclean target to the makefiles under manual

shindere committed:
Add the distclean target to the makefiles under otherlibs

shindere committed:
Slightly fix and clarify comment in runtime/Makefile

shindere committed:
Add the distclean target to api_docgen/Makefile

dra27 commented on Makefile:
```diff
@@ -1154,9 +1154,17 @@ depend: beforedepend
 	$(MAKE) -C stdlib distclean
+	$(MAKE) -C testsuite distclean
+	$(MAKE) -C tools distclean
+	$(MAKE) -C yacc distclean
```
This could be done with a pattern rule, I think? Something along the lines of:

```make
.PHONY: distclean-%
distclean-%: clean
	$(MAKE) -C $* distclean

.PHONY: distclean
distclean: distclean-debugger distclean-manual distclean-ocaml # etc. not including clean
	rm -f boot/ocamlrun # ...
```
the benefit of doing that is that `make -j distclean` starts to mean something.

dra27 commented on Makefile:
```diff
@@ -155,7 +155,7 @@ boot/ocamlruns$(EXE):
 #   - Build flexlink and FlexDLL support objects
 #   - Build the runtime
-# runtime/ocamlrun is then installed to boot/ocamlrun and the stdlib artefacts
+# runtime/ocamlrun is then installed to boot/ocamlrun and the stdlib artifacts
```
Bloomin' US spellings ;)

gasche commented:
Do you also clean the testsuite? I remember getting bitten from time to time by the fact that `testsuite/tools` is not clean when I expect it to.

dra27 commented on Makefile:
```diff
@@ -123,7 +123,7 @@ configure: configure.ac aclocal.m4 build-aux/ocaml_version.m4 tools/autogen
 .PHONY: partialclean
 partialclean::
-	rm -f utils/config.ml utils/domainstate.ml utils/domainstate.mli
+	$(MAKE) -C utils partialclean
```
This is the only one I disagree with - `utils/Makefile` only exists for the Dune build - the `utils/` is not an independent unit in the way all the other directory `Makefile`s are.

dra27 commented on testsuite/Makefile:
```diff
@@ -335,6 +335,10 @@ clean:
 
+.PHONY: distclean
+distclean: clean
+	rm -f _log*
```
This needs to do `$(MAKE) -C tools distclean` and `$(MAKE) -C lib distclean`, I think?

shindere commented:
Gabriel Scherer (2021/11/10 11:31 -0800):
> Do you also clean the testsuite? I remember getting bitten from time
> to time by the fact that `testsuite/tools` is not clean when I expect
> it to.

Yes `make distclean` is supposed to clean it.

I'll try whether `make clean` also does and if not, I can add that to
the PR although it is not its primary purpose.

dra27 commented on api_docgen/Makefile:
```diff
@@ -27,5 +27,9 @@ odoc-%:
 	rm -rf build odoc/build ocamldoc/build
+
+.PHONY: distclean
+distclean: clean
```
I've not 100% convinced myself that these shouldn't be empty recipes (either with a single `@` command below or ending with a semi-colon), but I think it's OK for a PHONY target.

dra27 reviewed (commented):
I think it's a good change (apart from disagreeing about adding this to `utils/Makefile`, as that file isn't meant to exist).

I don't think we we can trivially move this to `Makefile.common` because `Makefile.common` is included before the default target is set?

shindere commented:
Gabriel Scherer (2021/11/10 11:31 -0800):
> Do you also clean the testsuite? I remember getting bitten from time
> to time by the fact that `testsuite/tools` is not clean when I expect
> it to.

I couldn't reproduce. Running `make clean` in the testsuite removes
everything except `_log`. Plus, running `make clean` from the root
directory does clean up the testsuite.

It would bbe awesome and helpful if you could give steps to reproduce.

shindere committed:
Add the distclean target to the makefiles in the testsuite

shindere commented:
David Allsopp (2021/11/10 11:48 -0800):
> @dra27 commented on this pull request.
>
> I think it's a good change (apart from disagreeing about adding this to `utils/Makefile`, as that file isn't meant to exist).
>
> I don't think we we can trivially move this to `Makefile.common`
> because `Makefile.common` is included before the default target is
> set?

Ah yes, I think you are right. I didn't even think of that one. Thanks.

> >  	$(MAKE) -C runtime distclean
>  	$(MAKE) -C stdlib distclean
> +	$(MAKE) -C testsuite distclean
> +	$(MAKE) -C tools distclean
> +	$(MAKE) -C yacc distclean
>
> This could be done with a pattern rule, I think? Something along the lines of:
>
> ```make
> .PHONY: distclean-%
> distclean-%: clean
> 	$(MAKE) -C $* distclean
>
> .PHONY: distclean
> distclean: distclean-debugger distclean-manual distclean-ocaml # etc. not including clean
> 	rm -f boot/ocamlrun # ...
> ```
> the benefit of doing that is that `make -j distclean` starts to mean
> something.

Ah interesting. Thanks. Will give that a try next week.

> > @@ -155,7 +155,7 @@ boot/ocamlruns$(EXE):
>  #   - Build the standard library using boot/ocamlruns
>  #   - Build flexlink and FlexDLL support objects
>  #   - Build the runtime
> -# runtime/ocamlrun is then installed to boot/ocamlrun and the stdlib artefacts
> +# runtime/ocamlrun is then installed to boot/ocamlrun and the stdlib artifacts
>
> Bloomin' US spellings ;)

I removed the commit htat was doing just that.

> > @@ -123,7 +123,7 @@ configure: configure.ac aclocal.m4 build-aux/ocaml_version.m4 tools/autogen
>
>  .PHONY: partialclean
>  partialclean::
> -	rm -f utils/config.ml utils/domainstate.ml utils/domainstate.mli
> +	$(MAKE) -C utils partialclean
>
> This is the only one I disagree with - `utils/Makefile` only exists
> for the Dune build - the `utils/` is not an independent unit in the
> way all the other directory `Makefile`s are.

Understood. I won't fight much for that but let me give just one
arguemnt for leaving this as it is. That way, `utils/Makefile` is
responsible for both creating and removing `config.ml`, which does not
seem totallly insane to me. S, how about living this as it is for as
long as this file exists? I think in not too long we will be able to
generate `utils/config.ml` from a `utils/config.ml.in` at configure so
the whole `utils/Makefile` can be deleted and won't be necessary any
longer even for Dune. what do you think?

> > @@ -335,6 +335,10 @@ clean:
>  	find . -name '*_ocamltest*' | xargs rm -rf
>  	rm -f $(failstamp)
>
> +.PHONY: distclean
> +distclean: clean
> +	rm -f _log*
>
> This needs to do `$(MAKE) -C tools distclean` and `$(MAKE) -C lib
> distclean`, I think?

Yes, indeed, thanks! Fixed.

> >  clean:
>  	rm -rf build odoc/build ocamldoc/build
> +
> +.PHONY: distclean
> +distclean: clean
>
> I've not 100% convinced myself that these shouldn't be empty recipes
> (either with a single `@` command below or ending with a semi-colon),
> but I think it's OK for a PHONY target.

Yes I think it works as it is. It's done that way in several places at
the moment.

dra27 commented:
For `utils/Makefile`, I'm likewise not _that_ bothered by it! The only counter is that it's deleting `utils/domainstate.ml*` which it didn't generate.

shindere commented:
David Allsopp (2021/11/11 03:08 -0800):
> For `utils/Makefile`, I'm likewise not _that_ bothered by it! The only
> counter is that it's deleting `utils/domainstate.ml*` which it didn't
> generate.

Fair enough!

I pushed an updated version of the branch that leaves `utils/Makefile`
alone. Thanks!

shindere commented:
David Allsopp (2021/11/10 11:48 -0800):
> This could be done with a pattern rule, I think? Something along the lines of:
>
> ```make
> .PHONY: distclean-%
> distclean-%: clean
> 	$(MAKE) -C $* distclean
>
> .PHONY: distclean
> distclean: distclean-debugger distclean-manual distclean-ocaml # etc. not including clean
> 	rm -f boot/ocamlrun # ...
> ```
> the benefit of doing that is that `make -j distclean` starts to mean
> something.

I still like the idea in principle, but given that `make clean` has not
been parallelised yet, is it that important to include the rewriting of
distclean to make it parallel here?

I'd say parallelising both targets (and perhaps install) may diserve a
PR on its own, especially if we try to make PRs as swallowable by our
fellow reviewers.

What do you think, @dra27?
